pyccl/baryons.py

```diff
@@ -0,0 +1,137 @@
+from . import ccllib as lib
+from .base import CCLAutoreprObject
+from .pyutils import check
+from .pk2d import Pk2D
+from .base import unlock_instance
+import numpy as np
+
+
+class BaryonicEffects(CCLAutoreprObject):
+    """`BaryonicEffect` obects are used to include the imprint of baryons
+    on the non-linear matter power spectrum. Their main ingredient is a
+    method `apply_baryons` that takes in a `ccl.Pk2D` and returns another
+    `ccl.Pk2D` object that now accounts for baryonic effects.
+    """
+
+    def apply_baryons(self, cosmo, pk2d, in_place=False):
```

Comment:
I'd rather keep `in_place`. `Pk2D` objects are memory-heavy, and I can see good uses for this (in fact, this is how we actually use BCM in CCL).

Reply:
I'm not worried about time efficiency. It's mostly about convenience and memory eficiency. There are many examples of libraries that allow in-place operations, and I don't see a strong reason not to do that here.

Reply:
I have already replied to the memory efficiency argument: memory efficiency is irrelevant because of the `gbc`.

As far as convenience is concerned, I'd say having in-place operations is error-prone and unnecessarily complicates the code. We introduced `CCLObject` to make objects immutable, and the point was to only use `@unlock_instance` where it is unavoidable. I can't stress the immutability argument enough. We should avoid this if we can.

Reply:
I'd rather keep this, as reasoned above. We do this kind of operation in several places in the library. I am definitely not as worried about object immutability as you are, especially once proper `__eq__`s are in place.

Reply:
We don't - we only do it where it can't be avoided: dunder methods of `Pk2D`, setting power spectra in `Cosmology`, getting ingredients in `HMCalculator`, and adding new tracers in `Tracer`. Each of these use-cases is special and can't be worked around.

Here, we can work around that. Why would we want to mutate this (and add more lines of code), if we can completely avoid it?